gateway/ln-gateway/src/lib.rs

```diff
@@ -1545,6 +1560,234 @@ pub(crate) async fn fetch_lightning_node_info(
     Ok((node_pub_key, alias, network))
 }
 
+impl Gateway {
+    async fn public_key(&self, federation_id: &FederationId) -> Option<PublicKey> {
+        self.clients.read().await.get(federation_id).map(|client| {
+            client
+                .value()
+                .get_first_module::<GatewayClientModuleV2>()
+                .keypair
+                .public_key()
+        })
+    }
+
+    pub async fn payment_info(&self, federation_id: &FederationId) -> Option<PaymentInfo> {
+        Some(PaymentInfo {
+            public_key: self.public_key(federation_id).await?,
+            payment_fees: self.payment_fees(),
+            outgoing_cltv_delta: OUTGOING_CLTV_DELTA_NG,
+        })
+    }
+
+    pub fn payment_fees(&self) -> PaymentFees {
+        PaymentFees {
+            // we take a fee of one percent for outgoing contracts
+            send: PaymentFee::default(),
+            // we take a fee of one percent for incoming contracts
+            receive: PaymentFee::default(),
+        }
+    }
+
+    async fn send_payment(
+        &self,
+        payload: SendPaymentPayload,
+    ) -> std::result::Result<std::result::Result<[u8; 32], Signature>, SendPaymentError> {
+        let clients = self.clients.read().await;
+
+        let client = clients
+            .get(&payload.federation_id)
+            .ok_or(SendPaymentError::UnknownFederationId)?
+            .value();
+
+        let operation_id = OperationId(payload.contract.contract_id().0.into_inner());
+
+        let module = client.get_first_module::<GatewayClientModuleV2>();
+
+        if client.operation_exists(operation_id).await {
+            return Ok(module.subscribe_send(operation_id, payload.contract).await);
+        }
+
+        if payload.contract.claim_pk != module.keypair.public_key() {
+            return Err(SendPaymentError::NotOurKey);
+        }
+
+        if payload.invoice.consensus_hash::<sha256::Hash>() != payload.contract.invoice_hash {
```

Comment:
Invoice should already have the payment hash, we shouldn't need to hash it again